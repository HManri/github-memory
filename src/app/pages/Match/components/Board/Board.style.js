import styled from 'styled-components';

export const BoardWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  overflow-y: auto;
`;

export const BoardGame = styled.div`
  flex: 1;
  display: grid;
  grid-template-columns: repeat(4, 100px);
  grid-auto-rows: minmax(120px, auto);
  justify-content: center;
  gap: ${(props) => props.theme.spacings.medium}px;
`;

export const BoardResults = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: ${(props) => props.theme.spacings.small}px;
  margin-top: ${(props) => props.theme.spacings.medium}px;
`;

export const BoardResultsText = styled.div`
  ${(props) => props.theme.text.h2};
`;

export const BoardResultsActions = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
`;
