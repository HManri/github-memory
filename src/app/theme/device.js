const size = {
  mobile: '425px',
  desktop: '768px',
};

export const device = {
  mobile: `(min-width: ${size.mobile})`,
  desktop: `(min-width: ${size.desktop})`,
};
