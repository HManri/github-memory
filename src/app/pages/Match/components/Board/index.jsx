import React, { useEffect, useState } from 'react';
import { array, func } from 'prop-types';

import { useTimer } from 'hooks/useTimer';
import { GAME_RESULT } from 'constants/gameResult';
import Card from 'components/Card';
import { BoardWrapper } from './Board.style';

const DEFAULT_TIMER = 2 * 60;

export default function Board({ cards, onFinishGame }) {
  const [timer, startTimer, stopTimer] = useTimer(DEFAULT_TIMER);
  const [gameState, setGameState] = useState();
  const [selectedCardId, setSelectedCardId] = useState(null);
  const [selectedCardIndex, setSelectedCardIndex] = useState(null);
  const [selectedPairCardIndex, setSelectedPairCardIndex] = useState(null);
  const [numberOfPairs, setNumberOfPairs] = useState(0);

  console.log('aaa', timer);

  const handleOnClickCard = (cardId, index) => () => {
    if (!selectedCardId) {
      // there is no selected card
      setSelectedCardId(cardId);
      setSelectedCardIndex(index);
      return;
    }

    if (selectedCardId === cardId && selectedCardIndex === index) {
      // selected card is the same we already had, so hide it
      setSelectedCardId(null);
      setSelectedCardIndex(null);
      return;
    }

    if (selectedCardId === cardId) {
      // selected card have the same id, but it's not the same, we have a match
      setGameState({ ...gameState, [cardId]: true });
      setNumberOfPairs(numberOfPairs + 1);
    }

    // otherwise, we don't have a match
    setSelectedPairCardIndex(index);
    setTimeout(() => {
      setSelectedCardId(null);
      setSelectedCardIndex(null);
      setSelectedPairCardIndex(null);
    }, 400);
  };

  useEffect(() => {
    setGameState(
      [...new Set(cards.map((eachCard) => eachCard.id))].reduce((acc, eachId) => {
        acc[eachId] = false;
        return acc;
      }, {}),
    );
    setNumberOfPairs(0);
    // startTimer();
  }, [cards, startTimer]);

  useEffect(() => {
    if (gameState && numberOfPairs === Object.keys(gameState).length) {
      onFinishGame(GAME_RESULT.WIN);
      stopTimer();
    }
  }, [gameState, numberOfPairs, onFinishGame, stopTimer]);

  useEffect(() => {
    return () => {
      stopTimer();
    };
  }, [stopTimer]);

  if (!gameState) {
    // TODO show some kind of placeholder
    return null;
  }

  return (
    <BoardWrapper>
      {cards.map((eachCard, index) => (
        <Card
          key={`card-${index}-${eachCard.id}`}
          testId={`${index}-${eachCard.id}`}
          image={eachCard.imageSrc}
          imageAlt={eachCard.imageAlt}
          active={index === selectedCardIndex || index === selectedPairCardIndex}
          disabled={gameState[eachCard.id]}
          onClick={handleOnClickCard(eachCard.id, index)}
        />
      ))}
    </BoardWrapper>
  );
}

Board.propTypes = {
  cards: array.isRequired,
  onFinishGame: func.isRequired,
};
