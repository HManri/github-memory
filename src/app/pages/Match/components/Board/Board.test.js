import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import theme from 'theme/DefaultTheme';
import Board from '.';

describe('Board', () => {
  test('should render correctly', () => {
    const cards = [
      {
        id: '13352',
        imageSrc: 'https://avatars.githubusercontent.com/u/13352?v=4',
        imageAlt: 'cpojer',
        cardId: '0-13352',
      },
      {
        id: '13352',
        imageSrc: 'https://avatars.githubusercontent.com/u/13352?v=4',
        imageAlt: 'cpojer',
        cardId: '1-13352',
      },
    ];

    const board = render(
      <ThemeProvider theme={theme}>
        <Board testId="board-test" cards={cards} onFinishGame={() => null} />
      </ThemeProvider>,
    );

    expect(board).toMatchSnapshot();
  });
});
