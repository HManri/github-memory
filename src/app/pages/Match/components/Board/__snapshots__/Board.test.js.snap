// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-hLBbgP ghUaCh"
        data-testid="board-test"
      >
        <div
          class="sc-eDvSVe dhMAaz"
          data-testid="board-test__board_game"
        >
          <div
            class="sc-bcXHqe gsvhOc"
            data-testid="card-0-13352"
          >
            <img
              alt="cpojer"
              src="https://avatars.githubusercontent.com/u/13352?v=4"
            />
            <div
              class="sc-gswNZR OyGZg"
              data-testid="card-0-13352__translucid_layer"
            />
          </div>
          <div
            class="sc-bcXHqe gsvhOc"
            data-testid="card-1-13352"
          >
            <img
              alt="cpojer"
              src="https://avatars.githubusercontent.com/u/13352?v=4"
            />
            <div
              class="sc-gswNZR OyGZg"
              data-testid="card-1-13352__translucid_layer"
            />
          </div>
        </div>
        <div
          class="sc-jSUZER jbhCQc"
          data-testid="board-test__board_results"
        >
          <div
            class="sc-gKPRtg gQBhsg"
            data-testid="board-test__board_results__timer"
          >
            <span>
              Time: 120 seconds
            </span>
          </div>
          <div
            class="sc-gKPRtg gQBhsg"
            data-testid="board-test__board_results__score"
          >
            <span>
              Score: 0
            </span>
          </div>
          <div
            class="sc-iBYQkv gXKaxS"
            data-testid="board-test__board_results__actions"
          >
            <button
              class="sc-dkrFOg kyrTdB"
              data-testid="board-test__board_results__actions__give_up_btn"
            >
              <div
                class="sc-gKPRtg gQBhsg"
              >
                Give up
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-hLBbgP ghUaCh"
      data-testid="board-test"
    >
      <div
        class="sc-eDvSVe dhMAaz"
        data-testid="board-test__board_game"
      >
        <div
          class="sc-bcXHqe gsvhOc"
          data-testid="card-0-13352"
        >
          <img
            alt="cpojer"
            src="https://avatars.githubusercontent.com/u/13352?v=4"
          />
          <div
            class="sc-gswNZR OyGZg"
            data-testid="card-0-13352__translucid_layer"
          />
        </div>
        <div
          class="sc-bcXHqe gsvhOc"
          data-testid="card-1-13352"
        >
          <img
            alt="cpojer"
            src="https://avatars.githubusercontent.com/u/13352?v=4"
          />
          <div
            class="sc-gswNZR OyGZg"
            data-testid="card-1-13352__translucid_layer"
          />
        </div>
      </div>
      <div
        class="sc-jSUZER jbhCQc"
        data-testid="board-test__board_results"
      >
        <div
          class="sc-gKPRtg gQBhsg"
          data-testid="board-test__board_results__timer"
        >
          <span>
            Time: 120 seconds
          </span>
        </div>
        <div
          class="sc-gKPRtg gQBhsg"
          data-testid="board-test__board_results__score"
        >
          <span>
            Score: 0
          </span>
        </div>
        <div
          class="sc-iBYQkv gXKaxS"
          data-testid="board-test__board_results__actions"
        >
          <button
            class="sc-dkrFOg kyrTdB"
            data-testid="board-test__board_results__actions__give_up_btn"
          >
            <div
              class="sc-gKPRtg gQBhsg"
            >
              Give up
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
